### PR TITLE
[Azure Search][Management Plane] New features in 2019-10-01-Preview API, new 2020-03-13 API

### PR DESCRIPTION
**Pull request by Careyjmac**
### Description
This PR adds new features to the 2019-10-01-Preview api version, namely the CRUD operations for sharedPrivateLinkResources, a new list API for privateEndpointConnections, and adding support for the privateLinkResources API that returns the available configurations for these two APIs for any given search service.

This PR also adds a new 2020-03-13 api version, which makes the new features in the 2019-10-01-Preview api generally available.  It also adds a new "publicNetworkAccess" property to the search service definition, removing the previous "networkRuleSet/endpointAccess" property.

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by Careyjmac** [comment]
I don't seem to have access to add a label to this PR (apparently you need to be a contributor to do so?)  Can someone add the WaitForARMFeedback label for me since this PR contains a new api-version?

**Comment by weidongxu-microsoft** [comment]
@Careyjmac 

You probably want to show/describe what is the difference from a previous version (especially those in 2020-03-13/search.json, what is the difference between say 2015 version), and whether it is same change on 2020-03-13/search.json and 2019-10-01-preview/search.json (so reviewer only need to review one of them).

**Comment by Careyjmac** [comment]
> @Careyjmac
> 
> You probably want to show/describe what is the difference from a previous version (especially those in 2020-03-13/search.json, what is the difference between say 2015 version), and whether it is same change on 2020-03-13/search.json and 2019-10-01-preview/search.json (so reviewer only need to review one of them).

I included all of the changes and differences in my initial PR description.  Is there a more explicit way I can/should do it?  Copying and pasting here for reference:

This PR adds new features to the 2019-10-01-Preview api version, namely the CRUD operations for sharedPrivateLinkResources, a new list API for privateEndpointConnections, and adding support for the privateLinkResources API that returns the available configurations for these two APIs for any given search service.

This PR also adds a new 2020-03-13 api version, which makes the new features in the 2019-10-01-Preview api generally available. It also adds a new "publicNetworkAccess" property to the search service definition, removing the previous "networkRuleSet/endpointAccess" property.

**Comment by arv100kri** [comment]
> @Careyjmac
> 
> You probably want to show/describe what is the difference from a previous version (especially those in 2020-03-13/search.json, what is the difference between say 2015 version), and whether it is same change on 2020-03-13/search.json and 2019-10-01-preview/search.json (so reviewer only need to review one of them).

Hello, in addition to @Careyjmac description, here's maybe a visual diff: https://www.diffchecker.com/YN4Gskvt

Here's my cheat sheet for reviewing this PR:

1. Review the `2019-10-01-Preview` swagger to see the things we have added. Please note that we are "ok" with adding the new properties as it won't impact clients (we have telemetry to show that, even though technically it's a breaking swagger change). This aligns with what we have deployed in our backend, per guidance from Azure networking folks.

2. Then check out the diff link above: the only delta between the "GA" 2020-03-13 and the "preview" 2019-10-01-Preview API versions are listed there. We had to make one more "breaking" change in the `networkRuleSet` definition, also per guidance from Azure networking folks.

Please let us know if we can make reviewing this PR any easier for you?

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: This is inaccurate. Simply say: "Deletes the shared private link resource from the Search service."

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Same, we don't actually disconnect the shared private link resource - we delete it...

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: This needs to be "_CreateOrUpdate", and please modify the description accordingly.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: modify appropriately. 200, 201 - Created is also possible

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Can you call out that we can also possibly return 409 Conflict under a different scenario.
This happens when someone tries to immediately create a shared private link resource with the same name after it has been deleted. This is because when we delete the shared private link resource ARM actually returns a 202, and we don't yet poll/provide a URI for customers to poll. We return immediately.

Usually, this only takes less than a minute -

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Let's rephrase this description as - "The list of resources that are on boarded to private link service, that are supported by Azure Cognitive Search"

Comment (arv100kri) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: And maybe add another line saying, "For a given API version, this represents the 'supported' groupIds when creating a shared private link resources"

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: let's expand on what "shareable" means in all the description fields succintly. Something like - 
shareable private link resource type <==> "resource type that has been onboarded to private link service, supported by Azure Cognitive Search."

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: let's remove "available", don't know if it adds much to the context

Comment (arv100kri) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Also from everywhere below

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: The resource provider type for the *resource* that has been on boarded to private link service, supported by Azure Cognitive Search.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: The resource provider group id for the *resource* that has been on boarded to private link service, supported by Azure Cognitive Search.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: *within* the specified resource group?

File: specification/search/resource-manager/Microsoft.Search/stable/2020-03-13/examples/UpdateSharedPrivateLinkResource.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/stable/2020-03-13/examples/UpdateSharedPrivateLinkResource.json]: needs to be "CreateOrUpdate"

File: specification/search/resource-manager/Microsoft.Search/stable/2020-03-13/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/stable/2020-03-13/search.json]: Please incorporate the same comments I made from `2019-10-01-Preview` into here :)
Also the examples :)

**Review by arv100kri — CHANGES_REQUESTED** [review]
Most comments are nitty, but significant ones are:
- "CreateOrUpdate" shared private link resource
- 409 Conflict on Create shared private link resource
- Remove the "disconnect" verbiage from the delete shared private link resource API
- Expand on "shareable"

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: This was copied from the one for privateEndpointConnections, so I was just trying to be consistent.  Should I update that one to be CreateOrUpdate too?

Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: I didn't update the other one, but I did update it here.

Comment (arv100kri) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Right, it's accurate for the "privateEndpointConnections", because the "Create" private endpoint into Azure Search goes to Network RP and not ours... only calls to "Update" come to us...

Whereas for shared private link resources, we handle both create and update

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: So sorry, super nit.... "when creating a shared private link *resource*"

**Review by arv100kri — APPROVED** [review]
I only could spot that one tiny error :) Other than that LGTM.
@abhi1509 / @huangbolun could one of you please sign off as SMEs

@weidongxu-microsoft , please wait until one of the above 2 sign off + we get sign off from ARM reviewers before we merge the PR :)

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (weidongxu-microsoft) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: You might want to consider x-ms-client-flatten (when nest properties inside properties)
https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-client-flatten

If it is shared from other resource provider then don't change.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
@arv100kri 
Thanks for the guide and visual diff. It really make this PR easier to review.

@Careyjmac 
Sorry, I've missed the fine print of description on 2020 stable version.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json
Comment (abhi1509) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json]: why do we have both 200 and 201 here?

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json
Comment (abhi1509) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json]: not sure if we internally standardize what message customer sees when we create the PE. In future potentially we can ask customers for what message they want to see on PE Approval request

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json]: Because @arv100kri requested it earlier in this PR to show both the create and update scenario.

**Comment by abhi1509** [comment]
how does version info on 
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/search/resource-manager/readme.md change? 
is that automatic or we need to manually update it

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json]: This is actually already part of the input of the request.  I can change the message we have in the example if you want, but customers can put whatever they want here and that's what we'll send off on their behalf.

Comment (abhi1509) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/examples/CreateOrUpdateSharedPrivateLinkResource.json]: cool. just checked the spec as well.

**Review by abhi1509 — APPROVED** [review]

**Comment by abhi1509** [comment]
> > how does version info on
> > https://github.com/Azure/azure-rest-api-specs/blob/master/specification/search/resource-manager/readme.md change?
> > is that automatic or we need to manually update it
> 
> So I added a couple of changes there to add the configuration for 2020-03-13, but there is potentially some more places where we only reference GA versions that I can add this version as well. I assume that is what is expected? I can do that.

https://github.com/Azure/azure-rest-api-specs/blob/92b3e818d5143caae63e240510b7b1411ae0f5db/specification/search/resource-manager/readme.md#basic-information
This is where we need to find if it needs to also change. Its needed for SDK generation to correctly find the right version

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: I think we're just gonna leave it, this particular nested properties is only an output (i.e. customers can't set the values, can only retrieve them) and it is based on a spec from the networking team.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (arv100kri) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: this should be one underscore only

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: >null [](start = 26, length = 4)

"nextLink"

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: >PrivateLinkResources_ListSupported [](start = 24, length = 34)

just PrivateLinkResources_List

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: >PrivateEndpointConnections_ListByService [](start = 24, length = 40)

just PrivateEndpointConnections_List

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: >null [](start = 26, length = 4)

"nextLink"

just adding paging support in swagger.. otherwise, adding support later requires new api version

Comment (sanjaiganesh) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: consider updating existing definitions as well

---
In reply to: [389217784](https://github.com/Azure/azure-rest-api-specs/pull/8572#discussion_r389217784) [](ancestors = 389217784)

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: it is not required to list all the error scenarios.. just default error response is fine.

---
In reply to: [387213896](https://github.com/Azure/azure-rest-api-specs/pull/8572#discussion_r387213896) [](ancestors = 387213896)

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: is this async ? 

        "x-ms-long-running-operation": true,
        "x-ms-long-running-operation-options": {
          "final-state-via": "azure-async-operation"  or "location" based on how async is communicated
        }

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: >null [](start = 26, length = 4)

"nextLink"

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: >} [](start = 8, length = 1)

add
    "nextLink": {
          "type": "string",
          "description": "The URL to get the next set of operations, if there are any."
        }

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: >  } [](start = 6, length = 3)

add
    "nextLink": {
          "type": "string",
          "description": "The URL to get the next set of operations, if there are any."
        }

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [inline specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: You should reference the common types to bring in 'id', 'name', 'type'...

[https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v1/types.json](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v1/types.json)


it has definitions for TrackedResource and ProxyResource

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (sanjaiganesh) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: esp in new api version

---
In reply to: [389217804](https://github.com/Azure/azure-rest-api-specs/pull/8572#discussion_r389217804) [](ancestors = 389217804,389217784)

**Comment by weidongxu-microsoft** [comment]
I see Sanjai left some comments for improvement.

When you have acted on them (or decide not to act), left me a comment and I will merge this PR.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: This particular one I left as is as we chose not to add it in the backend.  I presume this is because this return item isn't configurable by the user, and will only ever be a fixed list that we send, and we don't expect it to get so long that paging would be necessary/useful.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: It is not async so this is unnecessary.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Done in next commit

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Done in next commit

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: This particular one I left as is as we chose not to add it in the backend. I presume this is because this return item isn't configurable by the user, and will only ever be a fixed list that we send, and we don't expect it to get so long that paging would be necessary/useful.

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Hmmm, interesting.  It seems like we've written our own version of the Resource type, which is very similar to your definition for TrackedResource.  In theory we could also use the ProxyResource for this property.  However, I'm going to leave it as is for now and not risk any major breaking changes that could come with rewriting the swagger this way.  I'll create a work item for the team that properly owns this on our side to address it later (I'm just helping out with Jedi requirements since its been so crazy myself, so I don't think its my place to rewrite this using those definitions).

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Fixed in next commit

File: specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json
Comment (Careyjmac) [reply specification/search/resource-manager/Microsoft.Search/preview/2019-10-01-preview/search.json]: Work item # 641623 in our  back log.

**Comment by Careyjmac** [comment]
> I see Sanjai left some comments for improvement.
> 
> When you have acted on them (or decide not to act), left me a comment and I will merge this PR.

@weidongxu-microsoft I responded to the feedback and/or made follow up work items on our side.  Assuming the validations pass we should be good to go.  Thanks!

**Pull request merged**
